### PR TITLE
Add ProjectionTerm

### PR DESCRIPTION
**Pull request by JulianKnodt**
This starts adding terms into chalk so that associated consts can be equated.

### CODE REVIEW

**Comment by compiler-errors** [comment]
howdy @JulianKnodt, what's the status of this?

**Comment by JulianKnodt** [comment]
@compiler-errors Started on it, been caught up with some other deadlines so have not had a chance to progress, will get to it eventualllyyyy but I'm unsure when I'll be able to.

**Comment by HKalbasi** [comment]
Is this PR in the right direction? I think chalk should not get its hand dirty with const evaluation, since const eval needs handling and interpreting whole rust syntax, and doing that in chalk is against modularization. I imagine that the user of chalk should provide const eval infrastructure, and evaluate const generics so chalk has them in the concrete shape.

In this way, chalk doesn't need to normalize associated constants (it can't normalize completely without full const eval), it just needs to provide a function that computes an `ImplId + substitution ` from a `<Type as Trait>`, so that the caller can find associated constants and evaluate them (and other associated items for other reasons).

**Comment by JulianKnodt** [comment]
@HKalbasi It's been a minute since I started this, and it's a good reminder for me to go back and actually implement this (I'm starting to finally get more time on my hands), but the context for this is that it would make sense to support const equality in the context of trait bounds [(see here)](https://github.com/rust-lang/rust/issues/92827). While I don't know whether Chalk needs to support evaluating them, my understanding from a few months ago that chalk needs to support consts in a new position which it previously did not.

If any of this is wrong, please let me know, I haven't thought about this in a minute and should get back to it.

**Comment by HKalbasi** [comment]
I thought it is going to add a `NormalizeConst` (which is essentially const eval) similar to `NormalizeFn` in #726 and handles associated constants in `<Type as Trait>::ASSOC_CONST` position, but it is about associated constants in `Trait<ASSOC_CONST=?>` position, so my comment above is irrelevant. Sorry for noise.

**Comment by JulianKnodt** [comment]
@compiler-errors Do you know who I can ask about Chalk? I'm starting to run into things I am unsure about, and am unsure where to look.

**Comment by jackh726** [comment]
Generally I or @nikomatsakis should have the most knowledge of Chalk things. Easiest to ask in Zulip.

There is very little infrastructure for consts in Chalk. I think most/all of it is in `chalk-ir`, for variable kinds and such.

File: chalk-solve/src/clauses/program_clauses.rs [hunk 1]
```diff
@@ -123,13 +123,20 @@ impl<I: Interner> ToProgramClauses<I> for AssociatedTyValue<I> {
...
+                    builder.push_clause(
+                        Normalize {
+                            alias: AliasTy::Projection(projection.clone()),
+                            term: Term::Ty(ty),
+                        },
+                        impl_where_clauses.chain(assoc_term_where_clauses),
+                    );
+                }
+                AssociatedTermValueBound::Const(ct) => {
+                    todo!();
```
Comment (JulianKnodt) [inline chalk-solve/src/clauses/program_clauses.rs:137]: Am unsure why this is not causing anything to break

**Pull request closed without merging**